wasm/src/malloc.c

```diff
@@ -98,16 +134,95 @@ void opa_heap_top_set(unsigned int top)
     init_free();
 }
 
+// returns the free list applicable for the requested size.
+static struct heap_blocks * __opa_blocks(size_t size) {
+    for (int i = 0; i < sizeof(heap_free) / sizeof(struct heap_blocks); i++) {
+        struct heap_blocks *candidate = &heap_free[i];
+
+        if (size <= candidate->size || !candidate->fixed_size)
+        {
+            return candidate;
+        }
+    }
+
+    opa_abort("__opa_blocks: bug");
+    return NULL;
+}
+
 void *opa_malloc(size_t size)
 {
     init();
-    HEAP_CHECK();
 
     // Look for the first free block that is large enough. Split the found block if necessary.
 
-    for (struct heap_block *b = heap_free_start.next; b != &heap_free_end; b = b->next)
+    struct heap_blocks *blocks = __opa_blocks(size);
+    HEAP_CHECK(blocks);
+
+    struct heap_block *b = blocks->fixed_size ?
+        __opa_free_find_fixed(blocks) : __opa_free_find_varying(blocks, size);
+    if (b != NULL)
+    {
+        return b->data;
+    }
+
+    // Allocate a new block.
+
+    if (blocks->fixed_size)
+    {
+        size = blocks->size;
+    }
+
+    unsigned int ptr = heap_ptr;
+    size_t block_size = sizeof(struct heap_block) + size;
+    heap_ptr += block_size;
+
+    if (heap_ptr >= heap_top)
+    {
+        unsigned int pages = (block_size / WASM_PAGE_SIZE) + 1;
+        __builtin_wasm_memory_grow(0, pages);
+        heap_top += (pages * WASM_PAGE_SIZE);
+    }
+
+    b = (void *)ptr;
+    b->size = size;
+    b->prev = NULL;
+    b->next = NULL;
+
+    return b->data;
+}
+
+// returns a free block from the list, if available.
+static struct heap_block * __opa_free_find_fixed(struct heap_blocks *blocks)
+{
+    struct heap_block *start = &blocks->start;
+    struct heap_block *end = &blocks->end;
+    struct heap_block *b = start->next;
+
+    if (b != end)
+    {
+        b->prev->next = b->next;
+        b->next->prev = b->prev;
+        b->prev = NULL;
+        b->next = NULL;
+
+        HEAP_CHECK(blocks);
+
+        return b;
+    }
+
+    return NULL;
+}
+
+// finds a free block at least of given size, splitting the found block if the remaining block exceeds the minimum size.
+static struct heap_block * __opa_free_find_varying(struct heap_blocks *blocks, size_t size)
+{
+    struct heap_block *start = &blocks->start;
+    struct heap_block *end = &blocks->end;
+    size_t min_size = blocks->size;
+
+    for (struct heap_block *b = start->next; b != end; b = b->next)
     {
-        if (b->size > (sizeof(struct heap_block) + size))
+        if (b->size > (sizeof(struct heap_block) + min_size + size))
```

Comment:
If I'm understanding this correctly we don't support splitting any blocks smaller than 2*128+sizeof(heap_block), right? (the `size` had to be at least 128 to get here, and the `min_size` is 128) Do you think it would be worthwhile to allow splitting to smaller sizes and then move the remaining block to the appropriate list?

Reply:
Right, definitely, we could do something like that eventually. The only reason I opted here for not doing it because of keeping this patch as simple as possible...